test/feasibility_checker.jl

```diff
@@ -0,0 +1,234 @@
+    report = primal_feasibility_report(model, point)
+    @test report[BinaryRef(x)] ≈ 0.1
+    @test report[UpperBoundRef(y)] ≈ 0.1
+    @test report[IntegerRef(y)] ≈ 0.1
+    @test report[FixRef(z)] ≈ 0.5
+    @test length(report) == 4
+end
+
+function test_scalar_affine()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c1, x <= 0.5)
+    @constraint(model, c2, x >= 1.25)
+    @constraint(model, c3, x == 1.1)
+    @constraint(model, c4, 0 <= x <= 0.5)
+    report = primal_feasibility_report(model, Dict(x => 1.0))
+    @test report[c1] ≈ 0.5
+    @test report[c2] ≈ 0.25
+    @test report[c3] ≈ 0.1
+    @test report[c4] ≈ 0.5
+    @test length(report) == 4
+end
+
+function test_scalar_quadratic()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c1, x^2 + x <= 0.5)
+    @constraint(model, c2, x^2 - x >= 1.25)
+    @constraint(model, c3, x^2 + x == 1.1)
+    @constraint(model, c4, 0 <= x^2 + x <= 0.5)
+    report = primal_feasibility_report(model, Dict(x => 1.0))
+    @test report[c1] ≈ 1.5
+    @test report[c2] ≈ 1.25
+    @test report[c3] ≈ 0.9
+    @test report[c4] ≈ 1.5
+    @test length(report) == 4
+end
+
+function test_vector()
+    model = Model()
+    @variable(model, x[1:3])
+    @constraint(model, c1, x in MOI.Nonnegatives(3))
+    @constraint(model, c2, x in MOI.Nonpositives(3))
+    @constraint(model, c3, x in MOI.Reals(3))
+    @constraint(model, c4, x in MOI.Zeros(3))
+    point = Dict(x[1] => 1.0, x[2] => -1.0)
+    report = primal_feasibility_report(model, point)
+    @test report[c1] ≈ 1.0
+    @test report[c2] ≈ 1.0
+    @test !haskey(report, c3)
+    @test report[c4] ≈ sqrt(2)
+    @test length(report) == 3
+end
+
+function test_vector_affine()
+    model = Model()
+    @variable(model, x[1:3])
+    @constraint(model, c1, 2 * x in MOI.Nonnegatives(3))
+    @constraint(model, c2, 2 * x in MOI.Nonpositives(3))
+    @constraint(model, c3, 2 * x in MOI.Reals(3))
+    @constraint(model, c4, 2 * x in MOI.Zeros(3))
+    point = Dict(x[1] => 1.0, x[2] => -1.0)
+    report = primal_feasibility_report(model, point)
+    @test report[c1] ≈ 2.0
+    @test report[c2] ≈ 2.0
+    @test !haskey(report, c3)
+    @test report[c4] ≈ sqrt(8)
+    @test length(report) == 3
+end
+
+function test_nonlinear()
+    model = Model()
+    @variable(model, x)
+    @NLconstraint(model, c1, sin(x) <= 0.0)
+    @NLconstraint(model, c2, sin(x) <= 1.0)
+    @NLconstraint(model, c3, exp(x) >= 2.0)
+    @NLconstraint(model, c4, x + x^2 - x^3 == 0.5)
+    report = primal_feasibility_report(model, Dict(x => 0.5))
+    @test report[c1] ≈ sin(0.5)
+    @test !haskey(report, c2)
+    @test report[c3] ≈ 2 - exp(0.5)
+    @test report[c4] ≈ 0.125
+end
+
+function runtests()
+    for name in names(@__MODULE__; all = true)
+        if !startswith("$(name)", "test_")
+            continue
+        end
+        @testset "$(name)" begin
+            getfield(@__MODULE__, name)()
+        end
+    end
```

Comment:
```suggestion
        if startswith("$(name)", "test_")
            @testset "$(name)" begin
                getfield(@__MODULE__, name)()
            end
        end
    end
```
the other block is not used

Reply:
In these situations I prefer early returns: https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

Reply:
I tend to do too but in that case there is just no "rest body" to early terminate, it's just one statement